src/platform/nrfconnect/FactoryDataParser.c

```diff
@@ -0,0 +1,160 @@
+                factoryData->date_month = 10 * (date.value[5] - '0') + date.value[6] - '0';
+                factoryData->date_day   = 10 * (date.value[8] - '0') + date.value[9] - '0';
+            }
+            else
+            {
+                LOG_ERR("Parsing error - wrong date format");
+            }
+        }
+        else if (strncmp("hw_ver_str", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->hw_ver_str);
+        }
+        else if (strncmp("rd_uid", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->rd_uid);
+        }
+        else if (strncmp("dac_cert", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->dac_cert);
+        }
+        else if (strncmp("dac_key", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->dac_priv_key);
+        }
+        else if (strncmp("pai_cert", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->pai_cert);
+        }
+        else if (strncmp("spake2_salt", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->spake2_salt);
+        }
+        else if (strncmp("spake2_verifier", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->spake2_verifier);
+        }
+        else if (strncmp("vendor_name", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->vendor_name);
+        }
+        else if (strncmp("product_name", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->product_name);
+        }
+        else if (strncmp("user", (const char *) currentString.value, currentString.len) == 0)
+        {
+            res = res && zcbor_bstr_decode(states, (struct zcbor_string *) &factoryData->user);
+        }
+        else
+        {
+            keyNotFound = true;
+            res         = res && zcbor_any_skip(states, NULL);
+        }
+
+        if (!keyNotFound && res)
+        {
+            elementsCount++;
+        }
+    }
+
+    res = zcbor_list_map_end_force_decode(states);
+
+    return res && elementsCount >= MIN_FACTORY_DATA_ELEMENTS;
```

Comment:
There seems no guarantee which elements are actually present.

Could we add some form of `FactoryData::IsValid` that checks for some rules (e.g. required strings not being nullptr)?

Reply:
That is kind of true, but some checks are already in place: https://github.com/project-chip/connectedhomeip/pull/19614/files#diff-02f990e612e48487589407830a65d53cde25ef741524b2629fdd30ca4ba0905dR136. I moved the responsibility for checking whether the values are present or not to the `FactoryDataProvider` as this is cpp object and it was much easier for me than making it in the C parser.

Reply:
It is much more obvious if the class itself can validate things.